src/Client.php

```diff
@@ -290,6 +290,153 @@ public function upload(string $path, $contents, $mode = 'add'): array
+     * @param bool   $close
+     *
+     * @return UploadSessionCursor
+     */
+    public function uploadSessionStart($contents, bool $close = false): UploadSessionCursor
+    {
+        $arguments = compact('close');
+
+        $response = json_decode(
+            $this->contentEndpointRequest('files/upload_session/start', $arguments, $contents)->getBody(),
+            true
+        );
+
+        return new UploadSessionCursor($response['session_id'], strlen($contents));
+    }
+
+    /**
+     * Append more data to an upload session.
+     * When the parameter close is set, this call will close the session.
+     * A single request should not upload more than 150 MB.
+     *
+     * @link https://www.dropbox.com/developers/documentation/http/documentation#files-upload_session-append_v2
+     *
+     * @param string              $contents
+     * @param UploadSessionCursor $cursor
+     * @param bool                $close
+     *
+     * @return \Spatie\Dropbox\UploadSessionCursor
+     */
+    public function uploadSessionAppend($contents, UploadSessionCursor $cursor, bool $close = false): UploadSessionCursor
+    {
+        $arguments = compact('cursor', 'close');
+
+        $this->contentEndpointRequest('files/upload_session/append_v2', $arguments, $contents);
+
+        $cursor->offset += strlen($contents);
+
+        return $cursor;
+    }
+
+    /**
+     * Finish an upload session and save the uploaded data to the given file path.
+     * A single request should not upload more than 150 MB.
+     *
+     * @link https://www.dropbox.com/developers/documentation/http/documentation#files-upload_session-finish
+     *
+     * @param string                              $contents
+     * @param \Spatie\Dropbox\UploadSessionCursor $cursor
+     * @param string                              $path
+     * @param string|array                        $mode
+     * @param bool                                $autorename
+     * @param bool                                $mute
+     *
+     * @return array
+     */
+    public function uploadSessionFinish($contents, UploadSessionCursor $cursor, string $path, $mode = 'add', $autorename = false, $mute = false): array
+    {
+        $arguments = compact('cursor');
+        $arguments['commit'] = compact('path', 'mode', 'autorename', 'mute');
+
+        $response = $this->contentEndpointRequest('files/upload_session/finish', $arguments, $contents);
+
+        return json_decode($response->getBody(), true);
+    }
+
+    /**
+     * Sometimes fread() returns less than the request number of bytes (for example, when reading
+     * from network streams).  This function repeatedly calls fread until the requested number of
+     * bytes have been read or we've reached EOF.
+     *
+     * @param resource $inStream
+     * @param int      $numBytes
+     * @throws Exception
+     * @return string
+     */
+    private static function readFully($inStream, int $numBytes)
+    {
+        $full = '';
```

Comment:
This method was copied from original `dropbox-sdk-php`.

I think that makes more sense calling it `readChunk` since it gets a chunk from the stream.